File: Backtracking/17_LetterCombinationsOfAPhoneNumber.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if len(digits) == 0:
            return []

        mp = {"2": "abc", "3": "def", "4": "ghi", "5": "jkl",
              "6": "mno", "7": "pqrs", "8": "tuv", "9": "wxyz"}

        def backtrack(index, path):
            if len(path) == len(digits):
                res.append("".join(path))
                return

            for letter in mp[digits[index]]:
                path.append(letter)
                backtrack(index + 1, path)
                path.pop()

        res = []
        backtrack(0, [])
        return res

File: Java Solutions/17_Letter Combinations of a Phone Number.java
class Solution {
    private List<String> res = new ArrayList<>();
    private Map<Character, String> letters = Map.of(
        '2', "abc",
        '3', "def",
        '4', "ghi",
        '5', "jkl",
        '6', "mno",
        '7', "pqrs",
        '8', "tuv",
        '9', "wxyz"
    );
    private String phoneDigits;

    public List<String> letterCombinations(String digits) {
        if (digits.length() == 0) {
            return res;
        }

        phoneDigits = digits;
        backtrack(0, new StringBuilder());
        return res;
    }

    private void backtrack(int i, StringBuilder path) {
        if (path.length() == phoneDigits.length()) {
            res.add(path.toString());
            return;
        }

        String options = letters.get(phoneDigits.charAt(i));
        for (char letter : options.toCharArray()) {
            path.append(letter);
            backtrack(i + 1, path);
            path.deleteCharAt(path.length() - 1);
        }
    }
}